src/daemon/rpmostreed-transaction-livefs.c

```diff
@@ -628,29 +699,64 @@ livefs_transaction_execute_inner (LiveFsTransaction *self,
-    if (!ostree_repo_checkout_at (repo, &rpmdb_checkout_opts, deployment_dfd, orig_rpmdb_path,
-                                  target_csum, cancellable, error))
+  /* And files that we always need to replace; the rpmdb, /usr/lib/passwd. We
+   * make a tmpdir just for this since it's a more convenient place to put
+   * temporary files/dirs without generating tempnames.
+   */
+  { g_auto(GLnxTmpDir) replace_tmpdir = { 0, };
+    if (!glnx_mkdtempat (ostree_repo_get_dfd (repo), "tmp/rpmostree-livefs.XXXXXX", 0700,
+                         &replace_tmpdir, error))
       return FALSE;
+
+    const char *replace_paths[] = { "/usr/share/rpm", "/usr/lib/passwd", "/usr/lib/group" };
+    for (guint i = 0; i < G_N_ELEMENTS(replace_paths); i++)
+      {
+        const char *replace_path = replace_paths[i];
+        if (!replace_subpath (repo, deployment_dfd, &replace_tmpdir,
+                              target_csum, replace_path,
+                              cancellable, error))
+          return FALSE;
+      }
   }
-  /* Now, RENAME_EXCHANGE the two */
-  if (glnx_renameat2_exchange (deployment_dfd, "usr/share/rpm", deployment_dfd, orig_rpmdb_path) < 0)
-    return glnx_throw_errno_prefix (error, "%s", "rename(..., RENAME_EXCHANGE) for rpmdb");
-  /* And nuke the old one */
-  if (!glnx_shutil_rm_rf_at (deployment_dfd, orig_rpmdb_path, cancellable, error))
-    return FALSE;
 
   rpmostree_output_task_end ("done");
 
+  if (diff->n_tmpfilesd > 0)
+    {
+      const char *tmpfiles_prefixes[] = { "/run/", "/var/"};
+      const char *tmpfiles_argv[] = { "systemd-tmpfiles", "--create",
+                                      "--prefix", NULL, NULL };
+
+      rpmostree_output_task_begin ("Running systemd-tmpfiles for /run,/var");
+
+      for (guint i = 0; i < G_N_ELEMENTS (tmpfiles_prefixes); i++)
+        {
+          const char *prefix = tmpfiles_prefixes[i];
+          GLNX_AUTO_PREFIX_ERROR ("Executing systemd-tmpfiles", error);
+          tmpfiles_argv[G_N_ELEMENTS(tmpfiles_argv)-2] = prefix;
+          g_autoptr(GSubprocess) subproc = g_subprocess_newv ((const char *const*)tmpfiles_argv,
+                                                              G_SUBPROCESS_FLAGS_STDERR_PIPE |
+                                                              G_SUBPROCESS_FLAGS_STDOUT_SILENCE,
+                                                              error);
+          if (!subproc)
+            return FALSE;
+          g_autofree char *stderr_buf = NULL;
+          if (!g_subprocess_communicate_utf8 (subproc, NULL, cancellable,
+                                              NULL, &stderr_buf, error))
+            return FALSE;
+          if (!g_subprocess_get_successful (subproc))
+            {
+              /* Only dump stderr if it actually failed; otherwise we know
+               * there are (harmless) warnings, no need to bother everyone.
+               */
+              sd_journal_print (LOG_ERR, "systemd-tmpfiles failed: %s", stderr_buf);
```

Comment:
We should set an error too in this path, right?

Reply:
Yes!  Good catch.  Fixup ⬇️